rust/lance-encoding/src/encodings/physical/block_compress.rs

```diff
@@ -90,11 +89,65 @@ impl ZstdBufferCompressor {
     pub fn new(compression_level: i32) -> Self {
         Self { compression_level }
     }
+
+    // https://datatracker.ietf.org/doc/html/rfc8878
+    fn is_raw_stream_format(&self, input_buf: &[u8]) -> bool {
+        if input_buf.len() < 8 {
+            return true; // can't be length prefixed format if less than 8 bytes
+        }
+        // read the first 4 bytes as the magic number
+        let mut magic_buf = [0u8; 4];
+        magic_buf.copy_from_slice(&input_buf[..4]);
+        let magic = u32::from_le_bytes(magic_buf);
+
+        // see RFC 8878, section 3.1.1. Zstandard Frames, which defines the magic number
+        const ZSTD_MAGIC_NUMBER: u32 = 0xFD2FB528;
+        if magic == ZSTD_MAGIC_NUMBER {
+            // the compressed buffer starts like a Zstd frame.
+            // Per RFC 8878, the reserved bit (with Bit Number 3, the 4th bit) in the FHD (frame header descriptor) MUST be 0
+            // see section 3.1.1.1.1. 'Frame_Header_Descriptor' and section 3.1.1.1.1.4. 'Reserved Bit' for details
+            const FHD_BYTE_INDEX: usize = 4;
+            let fhd_byte = input_buf[FHD_BYTE_INDEX];
+            const FHD_RESERVED_BIT_MASK: u8 = 0b0001_0000;
+            let reserved_bit = fhd_byte & FHD_RESERVED_BIT_MASK;
+
+            if reserved_bit != 0 {
+                // this bit is 1. This is NOT a valid zstd frame.
+                // therefore, it must be length prefixed format where the length coincidentally
+                // started with the magic number
+                false
+            } else {
+                // the reserved bit is 0. This is consistent with a valid Zstd frame.
+                // treat it as raw stream format
+                true
+            }
+        } else {
+            // doesn't start with the magic number, so it can't be the raw stream format
+            false
+        }
+    }
+
+    fn decompress_length_prefixed_zstd(
+        &self,
+        input_buf: &[u8],
+        output_buf: &mut Vec<u8>,
+    ) -> Result<()> {
+        let mut len_buf = [0u8; 8];
+        len_buf.copy_from_slice(&input_buf[..8]);
+        let uncompressed_len = u64::from_le_bytes(len_buf) as usize;
+
+        let compressed_data = &input_buf[8..];
+        output_buf.resize(uncompressed_len, 0);
```

Comment:
Can you use `reserve` instead of `resize`?  Since this could be called multiple times with the same output buf.

Reply:
@westonpace 
Since zstd's `decompress_to_buffer` requires the output_buffer to be pre-allocated, I still use the `resize`.
However, I’ve added a `start` offset to ensure that this function (and the `compress` function) can be called multiple times with the same output buffer. A dedicated test case has been added to verify this usage.